### ISSUE TITLE
Keras model learns, but predictions fit past much better than future

### ISSUE THREAD

**Issue by Nestak2** [keras-team/keras#14357]
I am trying to reproduce a tensorflow official [tutorial](https://www.tensorflow.org/tutorials/structured_data/time_series?fbclid=IwAR1CfmX6adoEpeVF9hqc1eNMf7AJIZM0pEzWpyMvbfFizxsa2uR97yDvgKQ#recurrent_neural_network) for time series predictions, but using different data and different generators. In brief, I am using sliding windows of 7 days (data points) of meteorological data to predict the temperature for the 8th day. I keep the neural net architecture the same. I manage to make the net learn, it exhibits correct trends, but its predictions for day 8 fit much better the ground truth value for day 1 of the sliding window input, than they fit the ground truth of day 8. I see this on the figures and in the MSE results.

But in the official tensorflow tutorial the predictions were much better. What am I doing wrong? Thanks

Here is the code in a colab [notebook](https://colab.research.google.com/drive/16rnqY-nX0Krlh3p5bTkP3kmSVShZqWbo?usp=sharing)

Here is the meteo [data](https://drive.google.com/file/d/1H9luseDYGL_XDE44bqzzGIKZeamBV3ch/view?usp=sharing) (preprocessed, cleaned and normalized)

Here are the figures:

![download1](https://user-images.githubusercontent.com/16193553/104022216-ad6f6580-51c8-11eb-8761-3b1953b1db81.png)
![download2](https://user-images.githubusercontent.com/16193553/104022231-b2341980-51c8-11eb-89f6-d9bb8f25da27.png)

And here is the code, explicitly pasted:

```
import matplotlib.pyplot as plt
import numpy as np
from sklearn import preprocessing
import pandas as pd
import scipy

from tensorflow.keras.preprocessing.sequence import TimeseriesGenerator
from tensorflow.keras.models import Model, Sequential, load_model
from tensorflow.keras.optimizers import SGD, Adam
import tensorflow as tf
from tensorflow.keras.layers import *

logpath_ms = './best_model.h5'


# the data is already cleaned, preprocessed and normalized, but have a look at it if you want:
df0 = pd.read_csv('github_example_normalized.txt', index_col=0)
X = df0.values

"""use 7 days (data points) to predict the temperature for the folowing 3 days at each time step
creating a target array y that corresponds to predicting multiple days at once:
"""

y = np.stack((np.roll(X, 0, axis=0),
                np.roll(X, -1, axis=0),
                np.roll(X, -2, axis=0)),axis=1)

#only take the temperature feature from the data:
y = y[:,:,3] 

n_input = 7 #one input window contains 7 data points

# split the data into 80% training 10% validation, 10% testing
train_generator = TimeseriesGenerator(X[:(-n_input-1)], y[:(-n_input-1)], length=n_input, batch_size=8, end_index=int(0.8*len(X[:(-n_input-1)])))
val_generator = TimeseriesGenerator(X[:(-n_input-1)], y[:(-n_input-1)], length=n_input, batch_size=8, start_index=int(0.8*len(X[:(-n_input-1)])), end_index=int(0.9*len(X[:(-n_input-1)])))
test_generator = TimeseriesGenerator(X[:(-n_input-1)], y[:(-n_input-1)], length=n_input, batch_size=8, start_index=int(0.9*len(X[:(-n_input-1)])))

# with the below lines you can have a look how x and y look like
# for i in range(5):
#     x_, y_ = test_generator[i]
#     print(x_.shape)
#     print(y_.shape)
#     print('%s => %s' % (x_, y_))

modelsave_cb = tf.keras.callbacks.ModelCheckpoint(logpath_ms, monitor='val_loss', mode='min', verbose=1, save_best_only=True)

## model taken from here: https://www.tensorflow.org/tutorials/structured_data/time_series?fbclid=IwAR1CfmX6adoEpeVF9hqc1eNMf7AJIZM0pEzWpyMvbfFizxsa2uR97yDvgKQ#recurrent_neural_network
model = Sequential()
model.add(LSTM(32, return_sequences=True, input_shape=(n_input, 12)))
model.add(MaxPool1D(2))
model.add(Dense(1))
model.compile(loss='mean_squared_error', optimizer='adam', metrics=['mse'])

model.fit(train_generator, validation_data=val_generator, epochs=5, verbose=1,
             callbacks=[modelsave_cb])

model = load_model(logpath_ms)

y_pred = model.predict(test_generator)

"""Below an evaluation of the results: the NN learns the trend, but it learns more to follow values of 7 data points in the past, not to predict the future. And unfortunately if we compare to the MSE of a naive baseline, where the prediction for t1 is just the ground truth at t0 than the baseline scores better"""

fig = plt.figure(figsize=(14,7))
plt.plot(df_avg.index[test_generator.start_index:test_generator.end_index + 1], test_generator.targets[test_generator.start_index:test_generator.end_index + 1][:,0], label='ground truth')
plt.plot(df_avg.index[test_generator.start_index:test_generator.end_index + 1], y_pred[:,0,:], label='pred day 1 after sliding window')
plt.legend()

# like above, but predictions shifted 7 days into the past. Gives much better fit to the ground truth
fig = plt.figure(figsize=(14,7))
plt.plot(df_avg.index[test_generator.start_index:test_generator.end_index + 1], test_generator.targets[test_generator.start_index:test_generator.end_index + 1][:,0], label='ground truth')
plt.plot(df_avg.index[test_generator.start_index-7:test_generator.end_index + 1-7], y_pred[:,0,:], label='pred day 1, like above, shifted 7 data points into past')
plt.legend()

# calculating MSE of the predictions
print(sum(y_pred[:,0,:].flatten() - test_generator.targets[test_generator.start_index:test_generator.end_index + 1,0])**2/len(y_pred))

# calculating MSE of the predictions shifted by 7 points into the past
print(sum(y_pred[7:,0,:].flatten() - test_generator.targets[test_generator.start_index:test_generator.end_index + 1 - 7,0])**2/len(y_pred[:-7]))

# MSE of a naive baseline, if we just make the model predict a value at t1 that is equal to the value for the previous period t0
y_t1 = test_generator.targets[test_generator.start_index:test_generator.end_index + 1,0]
y_t0 = test_generator.targets[test_generator.start_index-1:test_generator.end_index,0]
print(sum(y_t1-y_t0)**2/len(y_pred))
```

**Comment by jvishnuvardhan** [comment]
@Nestak2 Sorry for the late response. Is this still an issue for you?

Can you check with recent TF/keras versions and share a simple standalone code (with the saved model) to reproduce the issue? Thanks!
